File: 101.py
def combinationSum(candidates, target):
    res = []

    def backtrack(i, path, total):
        if total == target:
            res.append(path[:])
            return
        if total > target or i == len(candidates):
            return

        # Pick the number
        path.append(candidates[i])
        backtrack(i, path, total + candidates[i])
        path.pop()

        # Skip the number
        backtrack(i + 1, path, total)

    backtrack(0, [], 0)
    return res

print(combinationSum([2,3,6,7], 7))

File: 101.js
function combinationSum(candidates, target) {
    const res = [];

    function backtrack(i, path, total) {
        if (total === target) {
            res.push([...path]);
            return;
        }
        if (total > target || i === candidates.length) return;

        // pick
        path.push(candidates[i]);
        backtrack(i, path, total + candidates[i]);
        path.pop();

        // skip
        backtrack(i + 1, path, total);
    }

    backtrack(0, [], 0);
    return res;
}

console.log(combinationSum([2, 3, 6, 7], 7));
